dlog/commitment/src/srs.rs

```diff
@@ -6,16 +6,22 @@ This source file implements the Marlin structured reference string primitive
 
 use crate::commitment::CommitmentCurve;
 pub use crate::{CommitmentField, QnrField};
-use ark_ff::{BigInteger, FromBytes, PrimeField, ToBytes};
+use rayon::prelude::*;
+use ark_ff::{Field, BigInteger, FromBytes, PrimeField, ToBytes};
+use ark_ec::{ProjectiveCurve, AffineCurve};
 use array_init::array_init;
 use blake2::{Blake2b, Digest};
 use groupmap::GroupMap;
 use std::io::{Read, Result as IoResult, Write};
+use std::collections::HashMap;
+use crate::fft::group_fft;
+use ark_poly::{EvaluationDomain, Radix2EvaluationDomain as D};
 
 #[derive(Debug, Clone)]
 pub struct SRS<G: CommitmentCurve> {
     pub g: Vec<G>, // for committing polynomials
     pub h: G,      // blinding
+    pub lagrange_bases: HashMap<usize, Vec<G>>, // Lagrange bases, per domain size
```

Comment:
btw, as these are all `pub`, each field can be commented with `///` comments and they'll appear in rustdoc